src/Sylius/Bundle/ApiBundle/Serializer/ProductDenormalizer.php

```diff
@@ -0,0 +1,170 @@
+use Symfony\Component\Serializer\Normalizer\AbstractNormalizer;
+use Symfony\Component\Serializer\Normalizer\ContextAwareDenormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareTrait;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductDenormalizer implements ContextAwareDenormalizerInterface, DenormalizerAwareInterface
+{
+    use DenormalizerAwareTrait;
+
+    private const ALREADY_CALLED = 'sylius_product_denormalizer_already_called';
+
+    public function __construct(
+        private IriConverterInterface $iriConverter,
+    ) {
+    }
+
+    public function supportsDenormalization(mixed $data, string $type, string $format = null, array $context = []): bool
+    {
+        return
+            !isset($context[self::ALREADY_CALLED]) &&
+            is_array($data) &&
+            is_a($type, ProductInterface::class, true)
+        ;
+    }
+
+    public function denormalize(mixed $data, string $type, string $format = null, array $context = [])
+    {
+        $context[self::ALREADY_CALLED] = true;
+        $data = (array) $data;
+
+        $data = $this->denormalizeOptions($data, $context);
+
+        $this->validateAttributes($data);
+        $data = $this->denormalizeAttributes($data);
+
+        return $this->denormalizer->denormalize($data, $type, $format, $context);
+    }
+
+    /**
+     * @param array<array-key, mixed> $data
+     * @param array<array-key, mixed> $context
+     *
+     * @return array<array-key, mixed>
+     */
+    private function denormalizeOptions(array $data, array $context): array
+    {
+        if (!isset($context[AbstractNormalizer::OBJECT_TO_POPULATE])) {
+            return $data;
+        }
+
+        if (!isset($data['options'])) {
+            return $data;
+        }
+
+        /** @var ProductInterface $product */
+        $product = $context[AbstractNormalizer::OBJECT_TO_POPULATE];
+        Assert::isInstanceOf($product, ProductInterface::class);
+
+        if (!$product->getVariants()->isEmpty()) {
+            unset($data['options']);
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param array<array-key, mixed> $data
+     *
+     * @return array<array-key, mixed>
+     */
+    private function denormalizeAttributes(array $data): array
+    {
+        if (!isset($data['attributes'])) {
+            return $data;
+        }
+
+        foreach ($data['attributes'] as $key => $attributeData) {
+            /** @var ProductAttributeInterface $attribute */
+            $attribute = $this->iriConverter->getResourceFromIri($attributeData['attribute']);
+
+            if (in_array($attribute->getType(), [DateAttributeType::TYPE, DateTimeAttributeType::TYPE], true)) {
+                $attributeData['value'] = new \DateTime($attributeData['value']);
+                $data['attributes'][$key] = $attributeData;
+            }
+        }
```

Comment:
Just wondering whether this cannot be delegated to a separate `ProductAttributeValueDenormalizer`, a similar case with the validation.
As it is now, I need to overwrite this whole class to add the denormalization of a custom attribute type.

Reply:
I agree that some of the logic could be extracted, by in order not to iterate any longer over this large PR, I would be in favour of refactor in a separate PR 🖖🏻

Reply:
And I've done that in this PR 😅